pymtl3/dsl/ComponentLevel3.py

```diff
@@ -97,6 +105,142 @@ def _construct( s ):
+    #   <FreeVarName2> = lambda_closure[<Idx2>].cell_contents
+    #   ...
+    #   <FreeVarNameN> = lambda_closure[<IdxN>].cell_contents
+    #   def _lambda__<lambda_blk_name>():
+    #     # the assignment statement appears here
+    #   return _lambda__<lambda_blk_name>
+
+    new_root = ast.Module( body=[
+      ast.FunctionDef(
+          name="closure",
+          args=ast.arguments(args=[ast.arg(arg="lambda_closure", annotation=None, lineno=1, col_offset=12)],
+                             vararg=None, kwonlyargs=[], kw_defaults=[], kwarg=None, defaults=[]),
+          body=[
+            ast.Assign(
+              targets=[ast.Name(id=var, ctx=ast.Store(), lineno=1+idx, col_offset=2)],
+              value=ast.Attribute(
+                value=ast.Subscript(
+                  value=ast.Name(
+                    id='lambda_closure',
+                    ctx=ast.Load(),
+                    lineno=1+idx, col_offset=5+len(var),
+                  ),
+                  slice=ast.Index(
+                    value=ast.Num(
+                      n=idx,
+                      lineno=1+idx, col_offset=19+len(var),
+                    ),
+                  ),
+                  ctx=ast.Load(),
+                  lineno=1+idx, col_offset=5+len(var),
+                ),
+                attr='cell_contents',
+                ctx=ast.Load(),
+                lineno=1+idx, col_offset=5+len(var),
+              ),
+              lineno=1+idx, col_offset=2,
+            ) for idx, var in enumerate(lamb.__code__.co_freevars)
+          ] + [ lambda_upblk ] + [
+            ast.Return(
+              value=ast.Name(
+                id=blk_name,
+                ctx=ast.Load(),
+                lineno=4+len(lamb.__code__.co_freevars), col_offset=9,
+              ),
+              lineno=4+len(lamb.__code__.co_freevars), col_offset=2,
+            )
+          ],
+          decorator_list=[],
+          returns=None,
+          lineno=1, col_offset=0,
+        )
+    ] )
+
+    # In Python 3 we need to supply a dict as local to get the newly
+    # compiled function from closure.
+    # Then `closure(lamb.__closure__)` returns the lambda update block with
+    # the correct free variables in its closure.
+
+    dict_local = {}
+    exec( compile(new_root, blk_name, "exec"), lamb.__globals__, dict_local )
+    blk = dict_local[ 'closure' ]( lamb.__closure__ )
+
+    # Add an attribute that provides info of the original lambda function.
+    # Might be useful for translation.
+
+    blk._pymtl_lambda_connection_info = {
+      "file_name" : inspect.getsourcefile( lamb ),
+      "line_no"   : line,
+      "line"      : srcs[0],
+    }
+
+    # Add the source code to linecache for the compiled function
+
+    new_src = "def {}():\n {}\n".format( blk_name, src.replace("//=", "=") )
+    linecache.cache[ blk_name ] = (len(new_src), None, new_src.splitlines(), blk_name)
+
+    ComponentLevel1.update( s, blk )
+    s._dsl.lambda_upblk_src[ blk ] = src
```

Comment:
@ptpan i guess if we change src to srcs[0], it will be the very original line of code